src/Evolution/Systems/Cce/AnalyticSolutions/WorldtubeData.hpp

```diff
@@ -0,0 +1,248 @@
+ * as spin-weighted derivatives over the sphere. Instead, it makes best sense to
+ * compute the global solution over the extraction sphere, and cache
+ * intermediate steps to avoid repeating potentially expensive tensor
+ * calculations.
+ */
+struct WorldtubeData : public PUP::able {
+  using creatable_classes = tmpl::list<>;
+
+  /// The set of available tags provided by the analytic solution
+  using tags = tmpl::list<
+      Tags::CauchyCartesianCoords, Tags::Dr<Tags::CauchyCartesianCoords>,
+      gr::Tags::SpacetimeMetric<3, ::Frame::Inertial, DataVector>,
+      ::Tags::dt<gr::Tags::SpacetimeMetric<3, ::Frame::Inertial, DataVector>>,
+      GeneralizedHarmonic::Tags::Pi<3, ::Frame::Inertial>,
+      GeneralizedHarmonic::Tags::Phi<3, ::Frame::Inertial>,
+      gr::Tags::SpatialMetric<3, ::Frame::Inertial, DataVector>,
+      ::Tags::dt<gr::Tags::SpatialMetric<3, ::Frame::Inertial, DataVector>>,
+      Tags::Dr<gr::Tags::SpatialMetric<3, ::Frame::Inertial, DataVector>>,
+      gr::Tags::Shift<3, ::Frame::Inertial, DataVector>,
+      ::Tags::dt<gr::Tags::Shift<3, ::Frame::Inertial, DataVector>>,
+      Tags::Dr<gr::Tags::Shift<3, ::Frame::Inertial, DataVector>>,
+      gr::Tags::Lapse<DataVector>, ::Tags::dt<gr::Tags::Lapse<DataVector>>,
+      Tags::Dr<gr::Tags::Lapse<DataVector>>, Tags::News>;
+
+  WRAPPED_PUPable_abstract(WorldtubeData);  // NOLINT
+
+  virtual std::unique_ptr<WorldtubeData> get_clone() const noexcept = 0;
+
+  /*!
+   * \brief Retrieve worldtube data represented by the analytic solution, at
+   * boundary angular resolution `l_max` and time `time`
+   *
+   * \details The set of requested tags are specified by the final argument,
+   * which must be a `tmpl::list` of tags to be retrieved. The set of available
+   * tags is found in `WorldtubeData::tags`, and includes coordinate and
+   * Jacobian quantities as well as metric quantities and derivatives thereof.
+   */
+  template <typename... Tags>
+  tuples::TaggedTuple<Tags...> variables(const size_t output_l_max,
+                                         const double time,
+                                         tmpl::list<Tags...> /*meta*/) const
+      noexcept {
+    prepare_solution(output_l_max, time);
+    return {cache_or_compute(output_l_max, time, Tags{})...};
+  }
+
+ protected:
+  template <typename Tag>
+  db::const_item_type<Tag>& cache_or_compute(size_t output_l_max, double time,
+                                             Tag /*meta*/) const noexcept {
+    auto& item_cache = get<IntermediateCacheTag<Tag>>(intermediate_cache_);
+    if (item_cache.l_max == output_l_max and item_cache.time == time) {
+      return item_cache.data;
+    }
+    auto& item = item_cache.data;
+    destructive_resize_components(
+        make_not_null(&item),
+        Spectral::Swsh::number_of_swsh_collocation_points(output_l_max));
+    variables_impl(make_not_null(&item), output_l_max, time, Tag{});
```

Comment:
Tag{} should be either `tmpl::list<Tag>` or `tmpl::type_<Tag>` to future-proof against tags having state